### PR TITLE
[bookmarks] Fix large and rotated thumbnail creation 

### LINKED ISSUE #24818: Bug when creating bookmark thumbnails
### Describe the bug
Here is a clear and concise description of what the problem is:

When creating bookmarks in a movie, if you have an advancedsettings.xml file containing the line...

`<imageres>9999</imageres>`

...your bookmark thumbnails get cut, only the upper left part of the image appears, like in here...

I have tested older versions of Kodi and the same happens, at least since Matrix (it's the oldest I tried), so it seems an old bug.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

The bookmark thumbnails should show fully, even if you use that line in your advancedsettings.xml, since according to the Kodi wiki...

`Library artwork
Note: Using <imageres>9999</imageres> and <fanartres>9999</fanartres> will allow caching artwork at original size. `

Whether you use that option or not, it should show this way...

## Actual Behavior

Already mentioned.

## Possible Fix

No idea.

### To Reproduce

1. Add the line `<imageres>9999</imageres>` to your advancedsettings.xml file
2. Run Kodi
3. Play a movie and create a new bookmark. The thumbnail will get cut as mentioned previously.

### Debuglog

The debuglog can be found here:

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [X] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 11
 - Kodi version: Matrix, Nexus, Omega (probably older versions too)

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

There were a few bugs in the bookmark thumbnail creation, all addressed in this PR:
1/ thumbnail clipped for size > screen size (original problem reported in #24818 - imageres=9999 is the extreme case, but happens at lower values too (ex. 2000 for 1080p screen)
2/ thumbnail of rotated videos distorted
3/ needlessly large thumbnails

The code could have been made a bit more compact at the expense of readability. I can change if requested (the two blocks of if(rotated) {} else {} could be compacted into one).

In more details:
1/Bookmark thumbnails are created by video rendering, unlike movie/episode thumbnails, created directly the source video.
It brings a constraint: rendering is clipped to screen size by scissors set by GraphicContext. Decoupling would be good in the future.
-> PR fix: thumbnail dimensions calculated to fit in screen size

2/ The video container rotation metadata flag was not taken into account, resulting wrong aspect ratio and distorted thumbnail image. The dimensions of the source are pre-rotation, rotation is applied when rendering
-> PR fix: switch width/height as appropriate when comparing source dimensions against screen size, rendering area, ...

3/ With the points above taken care of, the thumbnail of a DVD played at 4k resolution would create a 4k-sized thumbnail. This is inefficient. The thumbnail will be limited to the source size instead (taking into account non-square pixels - so not always source widthxsource height dimension)
-> PR fix: make sure thumbnail dimensions do not exceed source dimensions

To avoid making the code overly complex, picture adjustments applied during rendering (zoom, a/r adjustment, crop) are not taken into account.

## Motivation and context

Fix #24818 (set advanced settings "imageres" to 9999 for thumbs at source size)

Also noticed thumbs of rotated video were distorted.

## What is the effect on users?

Bookmark thumbs work for all sizes and orientation.

## Screenshots (if appropriate):
Before PR - imageres 9999 4k video 1080p screen > thumbnail 9999x5624
image

Before PR - imageres 1080 rotated DVD source 720x480 - thumbnail is 1080x810 (wrong ar)
image

Before PR - imageres 9999 rotated DVD source 720x480 - thumbnail is 9999x7499 (cumulates both issues)
image

With PR - imageres 9999 4k video 1080p screen
image

With PR - imageres 9999 rotated DVD source 720x480 - thumbnail is 540x720 (due to non-square pixels of source)
image

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.h [hunk 1]
```diff
@@ -85,6 +85,7 @@ class CBaseRenderer
   */
   void GetVideoRect(CRect& source, CRect& dest, CRect& view) const;
   float GetAspectRatio() const;
+  unsigned int GetOrientation() { return m_renderOrientation; }
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.h:88]: Could this method be `const`?

File: xbmc/cores/VideoPlayer/VideoPlayer.h [hunk 2]
```diff
@@ -340,6 +340,8 @@ class CVideoPlayer : public IPlayer, public CThread, public IVideoPlayer,
   void FlushRenderer() override;
   void SetRenderViewMode(int mode, float zoom, float par, float shift, bool stretch) override;
   float GetRenderAspectRatio() const override;
+  void GetRects(CRect & source, CRect& dest, CRect& view) const override;
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/VideoPlayer.h:343]: clang-format will complain I promise

File: xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp [hunk 3]
```diff
@@ -380,16 +378,58 @@ bool CGUIDialogVideoBookmarks::AddBookmark(CVideoInfoTag* tag)
...
+  // The advanced setting defines the max size of the largest dimension (depends on orientation)
+  const unsigned int maxThumbDim{
+      CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_imageRes};
+  unsigned int width{}, height{};
+
+  if (!rotated)
+  {
+    if (aspectRatio >= 1.0f)
+    {
+      width = std::min({maxThumbDim, viewWidth, renderWidth, srcWidth});
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp:409]: You should include `<algorithm>` to be sure `std::min`is avalaible.

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 4]
```diff
@@ -5034,6 +5034,16 @@ float CVideoPlayer::GetRenderAspectRatio() const
   return m_renderManager.GetAspectRatio();
 }
 
+void CVideoPlayer::GetRects(CRect & source, CRect& dest, CRect& view) const
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:5037]: clang-format will complain I promise

File: xbmc/application/ApplicationPlayer.h [hunk 5]
```diff
@@ -55,6 +55,8 @@ class CApplicationPlayer : public IApplicationComponent
   void FlushRenderer();
   void SetRenderViewMode(int mode, float zoom, float par, float shift, bool stretch);
   float GetRenderAspectRatio() const;
+  bool GetRects(CRect & source, CRect& dest, CRect& view);
```
Comment (ksooo) [inline xbmc/application/ApplicationPlayer.h:58]: clang-format will complain I promise

**Comment by fuzzard** [comment]
Jenkins build this please

File: xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.h (hunk 1 above)
Comment (CrystalP) [reply xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.h]: sure. done, along with the format fixes.

File: xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp (hunk 3 above)
Comment (CrystalP) [reply xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp:410]: done

**Comment by CrystalP** [comment]
Fixed another missing const, it should be good now.

Could be backported to Omega I think, it touches many files but for trivial changes.

**Comment by fuzzard** [comment]
Target a .1 release if you do backport please

**Comment by rmrector** [comment]
Good stuff.

Other video thumbnails (chapters and single 'thumb' image) are generated with [DVDFileInfo](https://github.com/xbmc/xbmc/blob/348313a4335d14be88c8c91e70bdb3a67a5d542b/xbmc/cores/VideoPlayer/DVDFileInfo.cpp#L90) - expanding that might look good for backporting. It's not coupled to the current video context and _may_ handle the initial cases correctly already.

File: xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp [hunk 6]
```diff
@@ -380,15 +378,37 @@ bool CGUIDialogVideoBookmarks::AddBookmark(CVideoInfoTag* tag)
...
+  const unsigned int srcHeight{static_cast<unsigned int>(srcRect.Height())};
+  const unsigned int renderWidth{static_cast<unsigned int>(renderRect.Width())};
+  const unsigned int renderHeight{static_cast<unsigned int>(renderRect.Height())};
+  const unsigned int viewWidth{static_cast<unsigned int>(viewRect.Width())};
+  const unsigned int viewHeight{static_cast<unsigned int>(viewRect.Height())};
+
+  if (!srcWidth || !srcHeight || !renderWidth || !renderHeight || !viewWidth || !viewHeight)
+    return false;
+
+  // FIXME: the renderer sets the scissors to the size of the screen (provided by graphiccontext),
```
Comment (sarbes) [inline xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp:398]: I don't think that the scissor is the limiting factor. I've skimmed through the code for the capture of a frame in GL. From what I can tell, we render to the default framebuffer, which is the limiting element. Scissors are just set to the FB size.

Even though we could render to a separate buffer (which we should if we want to keep the route), I think the current way is not optimal. Ideally, we would capture the original planes, and not the mangled output which might introduce scaling artifacts (amongst other problems). But that would require re-engineering a lot of the capture code.

tl;dr: scissors are not the issue, as far as I see.

Comment (CrystalP) [reply xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp:398]: Scissors are the limiting factor in Windows. A dedicated texture of the requested thumb size is allocated (ie could be 4k even if screen is 1080p) for render. The rendering is then clipped by the scissors to screen size, resulting in the reported issue. It was a bit tricky to catch when playing with the code.
Maybe you'd see something similar in GL if the target buffer was larger.
There are + and - to capture before or after render. Best would be to capture after any color/transfer conversion and deinterlacing, but before scaling.
There are also pathological videos with incorrect flags / ar that the user works around with video settings overrides. I ignored those in the PR, it makes things very complicated from outside of the renderer code.

Comment (sarbes) [reply xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp:398]: On GL, it doesn't seem to me that we render to a separate texture, so we are bound to the UI resolution. Maybe this could be problematic when the resolution is set to PAL/NTSC?

If I had the time, I would retool the capture to return the raw YUV frame, and let swscale do the rest. But that's just a remark, and not a problem for this PR.

File: xbmc/application/ApplicationPlayer.h [hunk 7]
```diff
@@ -55,6 +55,8 @@ class CApplicationPlayer : public IApplicationComponent
   void FlushRenderer();
   void SetRenderViewMode(int mode, float zoom, float par, float shift, bool stretch);
   float GetRenderAspectRatio() const;
+  bool GetRects(CRect& source, CRect& dest, CRect& view);
+  unsigned int GetOrientation();
```
Comment (ksooo) [inline xbmc/application/ApplicationPlayer.h:59]: Any reason these two methods not being `const`?

Comment (CrystalP) [reply xbmc/application/ApplicationPlayer.h]: no, there was a little bit of trickery I missed. done.

**Comment by rmrector** [comment]
> I'm not sure what you mean?

My idea was that you can add a parameter to that method for float time in seconds to use the same video thumbnail generation as the rest of Kodi. Could be too much change from current behavior for a backport, though.

**Comment by CrystalP** [comment]
I understand. It would be a simple code change, but a large behavior change and I can see a couple issues.
Performance: DVDFileInfo::ExtractThumbToTexture operates like a separate player, would reopen the file and then attempt to seek to the correct point, while videoplayer is going on the same file.
Accuracy: we may not be able to seek accurately enough as it's likely seek would only go to key frames (to be confirmed).

Something like what @sarbes wrote avoids those issues and is the better approach long term.

I didn't think that far, just wanted to get a non-risky fix out there.

**Comment by CrystalP** [comment]
I'll merge in a couple days if there are no objections.
